fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java

```diff
@@ -29,38 +29,80 @@
 import com.apple.foundationdb.record.logging.LogMessageKeys;
+import com.apple.foundationdb.record.metadata.Key;
+import com.apple.foundationdb.record.metadata.expressions.FieldKeyExpression;
+import com.apple.foundationdb.record.metadata.expressions.KeyExpression;
 import com.apple.foundationdb.record.provider.foundationdb.FDBIndexableRecord;
 import com.apple.foundationdb.record.provider.foundationdb.IndexMaintainerState;
 import com.apple.foundationdb.tuple.Tuple;
 import com.google.protobuf.ByteString;
-import com.google.protobuf.Descriptors;
 import com.google.protobuf.InvalidProtocolBufferException;
 import com.google.protobuf.Message;
 
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
-import java.util.Map;
+import java.util.Arrays;
+import java.util.List;
 import java.util.Objects;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
 
 import static com.apple.foundationdb.record.lucene.LuceneIndexOptions.INDEX_PARTITION_BY_TIMESTAMP;
+import static com.apple.foundationdb.record.metadata.Key.Expressions.field;
 import static com.apple.foundationdb.record.provider.foundationdb.FDBStoreTimer.Waits.WAIT_LOAD_LUCENE_PARTITION_METADATA;
 
 /**
  * Manage partitioning info for a <b>logical</b>, partitioned lucene index, in which each partition is a separate physical lucene index.
  */
 @API(API.Status.EXPERIMENTAL)
 public class LucenePartitioner {
+    private static final ConcurrentHashMap<String, KeyExpression> partitioningKeyExpressionCache = new ConcurrentHashMap<>();
     public static final int PARTITION_META_SUBSPACE = 0;
     public static final int PARTITION_DATA_SUBSPACE = 1;
     private final IndexMaintainerState state;
     private final boolean partitioningEnabled;
     private final String partitionTimestampFieldName;
+    private final KeyExpression partitioningKeyExpression;
 
     public LucenePartitioner(@Nonnull IndexMaintainerState state) {
         this.state = state;
         partitionTimestampFieldName = state.index.getOption(INDEX_PARTITION_BY_TIMESTAMP);
         this.partitioningEnabled = partitionTimestampFieldName != null;
+        if (partitioningEnabled && (partitionTimestampFieldName.isEmpty() || partitionTimestampFieldName.isBlank())) {
+            throw new RecordCoreArgumentException("Invalid partition timestamp field name", LogMessageKeys.FIELD_NAME, partitionTimestampFieldName);
+        }
+        this.partitioningKeyExpression = makePartitioningKeyExpression(partitionTimestampFieldName);
+    }
+
+    /**
+     * make (and cache) a key expression from the partitioning field name.
+     *
+     * @param partitionTimestampFieldName partitioning field name
+     * @return key expression
+     */
+    @Nullable
+    private KeyExpression makePartitioningKeyExpression(@Nullable final String partitionTimestampFieldName) {
+        if (partitionTimestampFieldName == null) {
```

Comment:
Because of where this is called, I don't think this will ever be true.

Reply:
If the index isn't partition, this will be `null`

Reply:
Yes, but I think if the index isn't partitioned, this method isn't called.

Reply:
I don't know, maybe you're seeing something I am not... :)